### PR TITLE
Fix quadratic blowup in Bun.markdown on unclosed link brackets

### PR DESCRIPTION
**Pull request by robobun**
### Problem

`Bun.markdown.html()` / `.ansi()` / `.react()` are quadratic on inputs with many unclosed or nested link brackets (found by parser fuzzing):

```
bun -e 'console.time("md"); Bun.markdown.html("[".repeat(40000)); console.timeEnd("md")'
```

5,000 `[` → ~95 ms · 10,000 → ~333 ms · 20,000 → ~1,280 ms · 40,000 → ~5,069 ms — doubling the input quadruples the time, and a 255 KB flood doesn't finish in 15+ seconds (`.ansi()` takes minutes). The same blowup exists for `(((([[[[…`, `> > > > [[[[…`, `[[[[…]]]]`, and the cmark pathological `[a](b…`, `[a](<b…`, `[ (](…`, `[](…`, and `[` + backtick-flood families. Anything rendering untrusted markdown (READMEs, chat messages, comments) is exposed to a trivially craftable DoS payload.

### Cause

The inline parser resolved every `[` candidate by scanning forward from that opener to find its matching `]` (`process_link`, `try_match_bracket_link` in `src/md/links.rs`), and it does this twice per paragraph (emphasis collection + rendering). n unclosed openers × O(n) scan each = O(n²). Several secondary per-candidate scans compound the same family:

- reference lookups normalized an O(n)-long label for every bracket (`[[[[…]]]]`)
- unclosed bare/angle destinations and `(…)` titles were rescanned to the end of the input per candidate (`[a](b`, `[a](<b`, `[ (](` repeated)
- unclosed backtick runs inside the bracket scan were re-measured once per backtick (`[` + `` ` ``×n)
- with wiki links enabled (always on for `.ansi()`), every `[[` candidate rescanned to the end of the line looking for `]]`

### Fix

- **Bracket-pair map**: one pass per inline slice records every `[` (outside code spans, HTML tags/autolinks, and escapes) together with its `]`; link processing looks the match up instead of rescanning. The map's backing vec lives on the `Parser` and is recycled across slices, the open-bracket stack is threaded through the pair entries themselves, and the builder SIMD-jumps between `[ ] \ ` <` (`bun_core::immutable::index_of_any` / `index_of_char`) instead of walking every byte — so typical documents do not allocate here and barely touch the text. Slices with no `]` at all (the literal `"[".repeat(n)` payload) skip the build entirely after two SIMD scans. The old forward scan remains only as a fallback for openers the map doesn't know about.
- **Reference lookups** (`lookup_ref_def`): return early when there are no definitions or the label exceeds the spec's 999-character link-label cap (the same cap `parse_ref_def` already applies when storing definitions).
- **Bare destinations**: cap parenthesis nesting at 32, matching cmark and commonmark.js (the spec explicitly allows implementations to impose this limit).
- **Angle destinations**: stop at an unescaped `<` (per spec) instead of scanning to the end of the input.
- **`(…)`-delimited titles**: reject an unescaped `(` (per spec) instead of scanning for the next `)` anywhere in the input.
- **Wiki links**: cap `[`/`]` nesting at 32 so the `]]` scan is bounded.

Visible behavior changes are limited to the spec-aligned edge cases above (>32-deep parens in a destination, unescaped `<` in an angle destination, unescaped `(` in a paren title, >32-deep brackets inside a wiki link) — each now matches cmark/commonmark.js or the spec wording. A 30K-case differential fuzz over bracket/paren/backtick/quote-heavy inputs renders byte-identically before/after, and the full markdown suite (CommonMark spec, GFM extensions, edge cases — 1030 tests) passes.

### Measurements

| input | before (1.4.0 release) | after (debug + ASAN build) |
|---|---|---|
| `"[".repeat(40000)` → `.html()` | 5.0 s | 23 ms |
| `"[".repeat(255000)` → `.html()` | minutes | 161 ms |
| `"[".repeat(1_000_000)` → `.html()` | — | 750 ms |
| `"[".repeat(255000)` → `.ansi()` | minutes | 377 ms |

Every previously-quadratic pattern in the set above now scales linearly (×2 input → ×~2 time); release numbers are another order of magnitude lower than the debug/ASAN column.

Not addressed here: `.ansi()` has a much milder, pre-existing superlinear cost on very long single lines of plain text (no brackets involved) coming from the terminal word-wrapper; it is unrelated to this family and orders of magnitude smaller.

### Tests

`test/js/bun/md/md-edge-cases.test.ts` (new "pathological bracket inputs" block):

- two child-process tests render the pathological corpus — unclosed/nested/blockquoted bracket floods, unclosed inline/angle destinations, unclosed paren titles, bracket + backtick flood, ref-def variant, plus `.ansi()` with the wiki-link flood — with a 30 s kill switch, so the old code fails fast instead of hanging the runner
- conformance tests lock in that long link text still renders as a link, destination paren nesting works at 3/32 deep and stops past the cap, unescaped `<` / `(` invalidate angle destinations / paren titles, 999-character reference labels still resolve while 1000-character ones don't, and wiki-link nesting is capped

### CODE REVIEW

**Comment by robobun** [comment]
<div><sup>Updated 4:55 PM PT - May 22nd, 2026</sup></div>

@robobun, your commit a2b762ba22a83be39a287cb2a79e972648d6aaea is building: [`#57032`](https://buildkite.com/bun/bun/builds/57032)

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Too many allocations. Find a cheaper way. Use the string index-ing functions that are SIMD.

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
Can you find a way to fix this without allocating more memory?

File: src/md/links.rs [hunk 1]
```diff
@@ -119,11 +273,14 @@ impl Parser<'_> {
...
             } else {
-                // Bare destination — balance parentheses
+                // Bare destination — balance parentheses (nesting depth is capped)
                 let mut paren_depth: u32 = 0;
                 while pos < content.len() && !helpers::is_whitespace(content[pos]) {
                     if content[pos] == b'(' {
                         paren_depth += 1;
+                        if paren_depth > MAX_LINK_DEST_PAREN_DEPTH {
+                            break;
+                        }
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/md/links.rs:351]: Fixed in 3edc0cb2fb — the over-cap destination now invalidates the inline-link attempt (`dest_valid`), so the offending `(` is never reparsed as a title opener, while the reference/shortcut fallback stays reachable like cmark: `[a](` + `(`×33 + `))` renders as literal text, and with `[a]: /x` defined it renders the shortcut link followed by the literal parens. `try_match_bracket_link` mirrors the same behavior so the emphasis pass agrees. Added the `(`×33 + `))` case to the paren-cap test.

File: test/js/bun/md/md-edge-cases.test.ts [hunk 2]
```diff
@@ -527,3 +528,114 @@ code
...
+          ["ref def then nested brackets", "[x]: /url\\n\\n" + fill(60000, "[") + fill(60000, "]"), out => out.includes("[[[") && out.includes("]]]")],
+        ];
+        for (const [name, input, check] of cases) {
+          const out = Bun.markdown.html(input);
+          if (!check(out)) throw new Error("unexpected output for " + name + ": " + JSON.stringify(out.slice(0, 200)));
+          console.log("OK " + name);
+        }
+        console.log("DONE");
+      `);
+  }, 90_000);
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/bun/md/md-edge-cases.test.ts:580]: Keeping the explicit timeout on just these two spawn tests: the children legitimately take several seconds under debug/ASAN even with the fix (they render ~1.5 MB of pathological input plus process startup), so bun test's default 5s per-test budget would flake locally, while the CI runner already passes a much larger `--timeout`. The 30s kill switch inside the child is what bounds the regression case.

**Comment by robobun** [comment]
Re the `src/bun_core/debug.rs` finding: that file was an accidental inclusion in d2c6366b (it had ended up in the local git index; it's a byte-identical copy of the file already on `main` from #31163, not new code). It was already removed before this review posted — d2c6366b was rewritten as 3edc0cb2fb and force-pushed, and the PR diff is back to only the five markdown files (`src/md/*.rs` + the test).

File: src/md/links.rs [hunk 3]
```diff
@@ -509,7 +763,7 @@ impl Parser<'_> {
                 // Skip images (![...]) — images are allowed inside links
                 let is_inner_image = pos > 0 && label[pos - 1] == b'!';
                 // Try to find matching ] and check for link syntax
-                let inner = self.try_match_bracket_link(label, pos);
+                let inner = self.try_match_bracket_link(label, pos, brackets, base);
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/md/links.rs:770]: Good catch — fixed in 19295fd2f7: the run is now skipped as a whole (same `pos += count` as the other code-span skips), and the trigger (`"[" + "\`".repeat(150000) + "[a](u)](v)"`) is in the pathological child-process corpus. On the current release that payload takes ~4 s at just 60 K backticks; it's ~13 ms for 150 K after the fix (debug+ASAN).

One intentional behavior note: this also makes the inner-link check agree with how code spans are actually paired. In `[``x`[a](u)`](v)` the single-backtick span covers `[a](u)`, so the outer construct is now a link (`<a href="v">``x<code>[a](u)</code></a>`), matching cmark — previously the label walk resynchronized inside the unclosed ``-run and demoted the outer link even though the renderer itself emitted `[a](u)` as code. That was the only output change across the 30 K-case differential fuzz; locked in with a conformance test.
